geode-lucene/src/main/java/org/apache/geode/cache/lucene/internal/distributed/LuceneQueryFunction.java

```diff
@@ -132,6 +133,25 @@ private LuceneIndexImpl getLuceneIndex(final Region region,
     try {
       index =
           (LuceneIndexImpl) service.getIndex(searchContext.getIndexName(), region.getFullPath());
+      if (index == null && service instanceof LuceneServiceImpl) {
+        if (((LuceneServiceImpl) service).getDefinedIndex(searchContext.getIndexName(),
+            region.getFullPath()) != null) {
+          // The node may be in the process of recovering, where we have the index defined but yet
+          // to be recovered
+          // If we retry fast enough, we could get a stack overflow based on the way function
+          // execution is currently written
+          // Instead we will add an artificial sleep to slow down the retry at this point
+          // Hopefully in the future, the function execution would retry without adding to the stack
+          // and this can be removed
+          try {
```

Comment:
Can there be a possibility that this node never recovers but another node is ready to execute the query and we are stuck here for 10 seconds?